### PR TITLE
Apply `get_metric_with_all_outputs` on train step

### PR DESCRIPTION
**Pull request by illian01**
## Description

Please include a summary in English, of the changes in this pull request. If it closes an issue, please mention it here.

Closes: #209 

You should link at least one existing issue for PR. Before your create a PR, please check to see if there is an issue for this change.  
PRs from forked repository not accepted.

## Change(s)

- Modify `get_metric_with_all_outputs` method of base pipeline and all tasks.
- Collect samples and call `get_metric_with_all_outputs` in train phase.

## Changelog

If you PR to `dev` branch, please add a brief summary of the change to the **Upcoming Release** section of the [`CHANGELOG.md`](https://github.com/Nota-NetsPresso/netspresso-trainer/blob/master/CHANGELOG.md) file and include a link to the PR (formatted in markdown) and a link to your github profile.

For example,

```
- Added a new feature by `@myusername` in [PR 2023](https://github.com/Nota-NetsPresso/netspresso-trainer/pull/2023)
```

## Code Formatting

If you PR to either `master` or `dev` branch, you should follow the code linting process. Please check your code with `lint_check.sh` in `./scripts` directory.
For more information, please read the contribution guide in `CONTRIBUTING.md`.

### CODE REVIEW

File: src/netspresso_trainer/pipelines/detection.py [hunk 1]
```diff
@@ -116,23 +116,27 @@ def test_step(self, batch):
...
-                pred_on_image['post_boxes'] = detection[..., :4]
-                pred_on_image['post_scores'] = detection[..., -1]
-                pred_on_image['post_labels'] = class_idx
-                pred.append(pred_on_image)
-        self.metric_factory.calc(pred, target=targets, phase='valid')
+    def get_metric_with_all_outputs(self, outputs, phase):
+        # TODO: Compute metrics for train phase
+        if phase == 'train':
+            return
+        else:
```
Comment (deepkyu) [inline src/netspresso_trainer/pipelines/detection.py:123]: Please erase unnecessary `else` (guard clause).

File: src/netspresso_trainer/pipelines/base.py [hunk 2]
```diff
@@ -130,7 +130,7 @@ def test_step(self, batch):
         raise NotImplementedError
 
     @abstractmethod
-    def get_metric_with_all_outputs(self, outputs):
+    def get_metric_with_all_outputs(self, outputs, phase):
```
Comment (deepkyu) [inline src/netspresso_trainer/pipelines/base.py:133]: Shall we type-hint with `Literal` if there is no assert check for `phase` string?

**Review by deepkyu — COMMENTED** [review]
Thanks for the revision. I added some comments :)

File: src/netspresso_trainer/pipelines/base.py (hunk 2 above)
Comment (illian01) [reply src/netspresso_trainer/pipelines/base.py]: Good. I added type-hint with `Literal` on `phase`.

**Review by deepkyu — APPROVED** [review]

**Pull request merged**